### PR TITLE
Implement ability to resize text

### LINKED ISSUE #73: Zoom
Hello,

I am using Caprine with Ubuntu 15.04. I think it would be nice to have the ability to zoom, as in scale the text/images. It would make reading the text much easier.

### PR DESCRIPTION
**Pull request by phamdaniel**
Fixes #73 

This is implemented in the same way as iMessages, in that instead of zooming in and out on the whole interface this will only affect the conversation's text size.

### CODE REVIEW

File: browser.js [hunk 1]
```diff
@@ -38,6 +39,24 @@ app.on('platform-theme-changed', () => {
 	setDarkMode();
 });
 
+ipc.on('zoom-reset', () => {
+	setZoom(1.0);
+});
+
+ipc.on('zoom-in', () => {
+	const zoomFactor = storage.get('zoomFactor') + 0.1;
+	if (zoomFactor <= 1.4) {
```
Comment (kevva) [inline browser.js]: Add extra line break between line 47 and 48.

File: browser.js [hunk 2]
```diff
@@ -38,6 +39,24 @@ app.on('platform-theme-changed', () => {
...
+		setZoom(zoomFactor);
+	}
+});
+
+ipc.on('zoom-out', () => {
+	const zoomFactor = storage.get('zoomFactor') - 0.1;
+	if (zoomFactor >= 0.7) {
```
Comment (kevva) [inline browser.js]: Same as above.

**Comment by phamdaniel** [comment]
Added in the line breaks

File: menu.js [hunk 3]
```diff
@@ -138,6 +138,32 @@ const darwinTpl = [
...
+			{
+				label: 'Make Text Normal Size',
+				accelerator: 'CmdOrCtrl+0',
+				click() {
+					sendAction('zoom-reset');
+				}
+			},
+			{
+				label: 'Make Text Bigger',
+				accelerator: 'CmdOrCtrl+=',
```
Comment (sindresorhus) [inline menu.js]: `=` => `+`

File: menu.js [hunk 4]
```diff
@@ -138,6 +138,32 @@ const darwinTpl = [
 		]
 	},
 	{
+		label: 'View',
+		submenu: [
```
Comment (sindresorhus) [inline menu.js]: This sub-menu is the same for all OSes, so just put it in a variable as reuse it.

File: menu.js [hunk 5]
```diff
@@ -138,6 +138,32 @@ const darwinTpl = [
...
+			{
+				label: 'Make Text Normal Size',
+				accelerator: 'CmdOrCtrl+0',
+				click() {
+					sendAction('zoom-reset');
+				}
+			},
+			{
+				label: 'Make Text Bigger',
```
Comment (sindresorhus) [inline menu.js]: > Increase Text Size

File: menu.js [hunk 6]
```diff
@@ -138,6 +138,32 @@ const darwinTpl = [
...
+				accelerator: 'CmdOrCtrl+=',
+				click() {
+					sendAction('zoom-in');
+				}
+			},
+			{
+				label: 'Make Text Smaller',
```
Comment (sindresorhus) [inline menu.js]: > Decrease Text Size

File: menu.js [hunk 7]
```diff
@@ -138,6 +138,32 @@ const darwinTpl = [
 		]
 	},
 	{
+		label: 'View',
+		submenu: [
+			{
+				label: 'Make Text Normal Size',
```
Comment (sindresorhus) [inline menu.js]: > Reset Text Size

File: browser.js [hunk 8]
```diff
@@ -75,6 +102,18 @@ if (process.platform === 'darwin') {
 // activate Dark Mode if it was set before quitting
 setDarkMode();
 
+// Inject a global style node to maintain zoom factor after conversation change.
+// Also set the zoom factor if it was set before quitting.
+document.addEventListener('DOMContentLoaded', () => {
+	const zoomFactor = storage.get('zoomFactor') || 1.0;
+	const style = document.createElement('style');
+	style.type = 'text/css';
```
Comment (sindresorhus) [inline browser.js]: This is the default, so no point in setting it.

File: browser.js [hunk 9]
```diff
@@ -38,6 +39,26 @@ app.on('platform-theme-changed', () => {
...
+
+	if (zoomFactor <= 1.4) {
+		setZoom(zoomFactor);
+	}
+});
+
+ipc.on('zoom-out', () => {
+	const zoomFactor = storage.get('zoomFactor') - 0.1;
+
+	if (zoomFactor >= 0.7) {
```
Comment (sindresorhus) [inline browser.js]: 0.8

File: browser.js [hunk 10]
```diff
@@ -38,6 +39,26 @@ app.on('platform-theme-changed', () => {
...
 });
 
+ipc.on('zoom-reset', () => {
+	setZoom(1.0);
+});
+
+ipc.on('zoom-in', () => {
+	const zoomFactor = storage.get('zoomFactor') + 0.1;
+
+	if (zoomFactor <= 1.4) {
```
Comment (sindresorhus) [inline browser.js]: 1.6

**Comment by sindresorhus** [comment]
Nice! Thanks for working on this @phamdaniel :)

File: browser.js [hunk 11]
```diff
@@ -67,6 +88,12 @@ function getNextIndex(next) {
 	return (index % list.length + list.length) % list.length;
 }
 
+function setZoom(zoomFactor) {
+	const node = document.getElementById('zoomFactor');
+	node.innerHTML = `${conversationSelector} {zoom: ${zoomFactor} !important}`;
```
Comment (sindresorhus) [inline browser.js]: `node.textContent` instead?

**Comment by phamdaniel** [comment]
It's always a pleasure getting feedback from you 😄

File: menu.js [hunk 12]
```diff
@@ -138,6 +138,9 @@ const darwinTpl = [
 		]
 	},
 	{
+		label: 'View'
```
Comment (sindresorhus) [inline menu.js]: I would just add the submenu here instead of at the bottom, for clarity:

```js
submenu: viewSubmenu
```

**Comment by phamdaniel** [comment]
Done! Though I had to move the the submenu declarations to the top of the file to get access to them.

File: menu.js [hunk 13]
```diff
@@ -16,6 +16,54 @@ function sendAction(action) {
...
+	{
+		label: 'Reset Text Size',
+		accelerator: 'CmdOrCtrl+0',
+		click() {
+			sendAction('zoom-reset');
+		}
+	},
+	{
+		label: 'Increase Text Size',
+		accelerator: 'CmdOrCtrl+Plus',
```
Comment (sindresorhus) [inline menu.js]: This shows up as

screen shot 2016-04-25 at 03 18 37

While it should be:

screen shot 2016-04-25 at 03 18 58


Notice the shortcut.

**Comment by sindresorhus** [comment]
Zooming no longer works.

**Comment by phamdaniel** [comment]
The shortcut problem is a bug in Electron. See https://github.com/electron/electron/issues/1507
As for the zooming not working, I'm not too sure... It's still working on my end.

**Comment by sindresorhus** [comment]
> The shortcut problem is a bug in Electron. See electron/electron#1507

Ah, I see. I went with `Cmd+=`. Then at least it works as intended when using the shortcuts.

> As for the zooming, I'm not too sure... It's still working on my end.

My mistake. I managed to use an older Electron version from another branch...

**Comment by sindresorhus** [comment]
Thank you! :) 🍻

**Pull request closed without merging**